feat: query profiling part 1: synchronous

Linchin opened the pull request:
The Synchronous section for query profiling support.

Checklist:
- [x] add the class definition query_profile.py
- [x] add class imports in init.py
- [x] methods:
  - [x] query.stream, .get
  - [x] collection.stream, .get  -- calls query
  - [ ] ~~CollectionGroup.stream, .get, -- inherits query~~
  - [x] AggregationQuery.stream, .get
  - [x] vector.stream, .get 
  - [x] transaction.get
  - [ ] ~~client.get_all~~
- [x] add system test for transaction
- [x] stream generator unit tests
- [x] DocumentSnapshotList class
- [x] DocumentSnapshotList unit tests
- [x] system test for each class
  - [x] aggregation
  - [x] collection
  - [ ] ~~document~~ only get, and does not use the stream api
  - [x] vector_query
  - [x] query
  - [x] transaction

Also closes #947

Linchin committed:
feat: support query profiling

Linchin committed:
collection

Linchin committed:
fix unit tests

Linchin committed:
unit tests

Linchin committed:
vector get and stream, unit tests

Linchin committed:
aggregation get and stream, unit tests

Linchin committed:
docstring

Linchin committed:
query profile unit tests

Linchin committed:
update base classes' method signature

Linchin committed:
documentsnapshotlist unit tests

Linchin committed:
func signatures

Linchin committed:
undo client.py change

Linchin committed:
transaction.get()

Linchin committed:
lint

Linchin committed:
system test

Linchin committed:
fix shim test

Linchin committed:
fix sys test

Linchin committed:
fix sys test

Linchin committed:
system test

Linchin committed:
another system test

Linchin committed:
skip system test in emulator

Linchin committed:
stream generator unit tests

Linchin committed:
coverage

Linchin committed:
add system tests

Linchin committed:
Merge branch 'main' into query-profiling-3

Linchin committed:
small fixes

Linchin committed:
undo document change

Linchin committed:
add system tests

Linchin committed:
vector query system tests

Linchin committed:
format

Linchin committed:
fix system test

Linchin committed:
comments

Linchin committed:
add system tests

Linchin committed:
improve stream generator

daniel-sanche reviewed (commented):
Leaving a couple initial thoughts, but still looking through things

daniel-sanche commented:
We would benefit from some time verification here, but the mypy check is [currently broken](https://github.com/googleapis/python-firestore/issues/773). Maybe you should try to enable it just for the new classes/code you touch? I don't know how practical that would be though

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -54,6 +58,7 @@ def get(
             retries.Retry, None, gapic_v1.method._MethodDefault
         ] = gapic_v1.method.DEFAULT,
         timeout: float | None = None,
+        explain_options: Optional[ExplainOptions] = None,
```
Also, probably better to follow existing style (from the timeout argument) and use `ExplainOptions | None` instead of `Optional`

Linchin replied:
> Can you add a * above `explain_options

Thanks for pointing this out, it's great to know! Added.

Linchin replied:
> Also, probably better to follow existing style (from the timeout argument) and use ExplainOptions | None instead of Optional

I'm not quite sure about this, because the type hint style is a mix right now. Maybe we can save it for a docstring overhaul later, when we are certain about which style we will go for?

daniel-sanche replied:
Yeah that works for me!

daniel-sanche commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -503,6 +504,48 @@ def _to_protobuf(self) -> Optional[Document]:
+
+    @property
+    def explain_options(self):
+        return self._explain_options
```
Does StreamGenerator also expose `explain_options`?

Linchin replied:
Good catch! Added the property.

Linchin commented on google/cloud/firestore_v1/base_transaction.py:
```diff
@@ -20,6 +20,9 @@
 from google.cloud.firestore_v1 import types
 
+# Types needed only for Type Hints
+from google.cloud.firestore_v1.query_profile import ExplainOptions
```
Indeed, I have added the type checking flag.

Linchin committed:
type checking

Linchin committed:
adding stars

Linchin commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -503,6 +504,48 @@ def _to_protobuf(self) -> Optional[Document]:
+    Args:
+        reference (:class:`~google.cloud.firestore_v1.document.DocumentReference`):
+            A document reference corresponding to the document that contains
+            the data in this snapshot.
```
yes, removed

Linchin committed:
delete comment

Linchin committed:
remove coverage requirements for type checking part

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -70,20 +75,37 @@ def get(
+            explain_metrics = None
+        else:
+            explain_metrics = result.explain_metrics
+
```
Indeed this same if/else code appears repeatedly in all get() methods. But considering we want to expand DocumentSnapshotList into a more general list class that may encompass types other than DocumentSnapshot, I wonder if the same logic checking explain_options will apply to all future classes that form the list? Also if we put this logic inside DocumentSnapshotList, we would need to pass the original `result` (StreamGenerator) into the initialization. I personally feel like it's cleaner to initialize with the explain_metrics, but I also want to hear your opinion on this.

daniel-sanche replied:
On second thought, I think you're right, it's probably better to leave most of this stuff outside the __init__

I was thinking it could be cleaner to just pass in a `StreamGenerator` on __init__, since it should contain all the information needed by `DocumentSnapshotList`. But that gets complicated on the async side, because __init__ isn't an async function. So we'd have trouble building the object there

Still though, one thing to consider: if we pass in the `StreamGenerator` and hold a reference to it in `DocumentSnapshotList`, you can delegate to that instance when calling `explain_metrics()`, which could help reduce some duplication

Linchin replied:
When I'm trying to implement this, I still feel a bit unsure. Passing `StreamGenerator` feels like we are coupling it with `QueryResultsList` really deeply and may make these two classes less versatile in the future.

daniel-sanche replied:
If it seems messy, don't worry about it, I was just throwing out ideas. We can just re-visit this if the duplicated code expands in the future

IMO I think it makes sense for `StreamGenerator` and `QueryResultsList` to be coupled in some way though, either through inheritance or composition. They have essentially the same use-case, QueryResultsList just has the stream fully loaded into memory

Linchin commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -503,6 +504,48 @@ def _to_protobuf(self) -> Optional[Document]:
+        if self._explain_options is None:
+            raise query_profile.QueryExplainError("explain_options not set on query.")
+        else:
+            return self._explain_metrics
```
I agree this is repeating code, but also I feel like using mixin for one property might be a bit of overkill? I think you have a good point that we are raising a lot of exceptions with exactly the same error message. Would it make sense to create a few subclasses of QueryExplainError for each of the cases, so at least we have a central place to update the error messages?

daniel-sanche replied:
Yeah, I think you're right, Mixins probably only make sense if the shared code starts to expand, so maybe just keep it in mind if it seems it may be useful at some point. Probably overkill for now

One other thought I mentioned in a different thread though: DocumentSnapshotList could hold an instance of StreamGenerator, and delegate this call to the inner stream instance, which would be another way to reduce duplication

Linchin replied:
Great idea! I will add it along with the other list related changes

Linchin committed:
add explain_options to StreamGenerator

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -38,3 +65,35 @@ def throw(self, exp=None):
+                next(self)
+            except StopIteration:
+                pass
+            return self._explain_metrics
```
there is a problem with the typing here: `self._explain_metrics` is optional, but this method says it always returns an `ExplainMetrics`. Maybe raise an exception if it's still None after calling next? Or is there a cleaner way?

Linchin replied:
Indeed, I think it's better to add an exception if `explain_metrics` is empty. I'm a bit curious as to when this can happen, maybe the backend didn't send the proper data?

Linchin replied:
done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -38,3 +65,35 @@ def throw(self, exp=None):
+        elif self._explain_options.analyze is False:
+            # we need to run the query to get the explain_metrics
+            try:
+                next(self)
```
This structure is going to cause issues when you make the async version. `@property` methods can't call coroutines.

Maybe we should just raise an exception if this is accessed before it's iterated? Or do you have another idea?

Linchin replied:
> Maybe we should just raise an exception if this is accessed before it's iterated? Or do you have another idea?

With `explain_options.analyze == False`, the query isn't really run, and there's no query result returned. Thus users won't organically iterate the results themselves. I'm not the most knowledgeable person about async python, but it seems possible to make an async property, but it's not recommended isn't very convenient to use. (see [stackoverflow thread](https://stackoverflow.com/questions/54984337/how-should-you-create-properties-when-using-asyncio).) 

Apart from forcing an async property, other options I can think of:
1. Move the logic to inside `Query._make_stream()` ([here](https://github.com/googleapis/python-firestore/blob/a0d2d7e26bd2008cc1f19727002a8c663bce78ec/google/cloud/firestore_v1/query.py#L362)): if `analyze == False`, get `explain_metrics` right after iterator is created, so StreamGenerator doesn't need to call the API anymore. This may make creating the generator a bit slower, though.
2. make explain_metrics not a property but a function. At least formally it's not a property anymore - but this will create a difference between datastore and firestore SDKs

I feel like (1) is a better idea. WDYT?

daniel-sanche replied:
I don't think 1 solves the problem. In that case, _make_stream will be an AsyncGenerator, and you'll need to await generator.__anext__ to get the explain_metrics value from it. (Or maybe I'm misunderstanding the details, but fundamentally, you'll have to await something somewhere to get that value, so we can't use a sync property)

I think 2 is cleaner. Turn it into `get_explain_metrics()`. I think it's ok that it's different from datastore here

Linchin replied:
For 1, I meant we get `explain_metrics` when  `Query._make_stream()` is called, and pass the value to the init function of StreamGenerator, so `explain_metrics` will be strictly a property without any operation to get data when called

daniel-sanche replied:
I don't think we can, because `self._make_stream` is just making an AsyncGenerator when it is [called for the first time](https://github.com/googleapis/python-firestore/blob/cb693f8b02434de803ce41e72bb5aa18c5de3764/google/cloud/firestore_v1/async_query.py#L401C32-L401C44). Basically like an __init__ function. It's code isn't executed until __anext__ is called on it.

If AsyncQuery.stream were an async function (aka coroutine), then we could do something there. But as-is, we're not in an async context at any point there

Linchin replied:
Wow I see, didn't know that `AsyncQuery.stream()` isn't async. Then I will go with option 2.

Linchin replied:
done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -14,21 +14,48 @@
 
 class StreamGenerator(abc.Generator):
-    """Generator for the streamed results."""
+    """Generator for the streamed results.
```
Please use generic for the typing here

Linchin replied:
done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -38,3 +65,35 @@ def throw(self, exp=None):
+        elif self._explain_options.analyze is False:
+            # we need to run the query to get the explain_metrics
+            try:
+                next(self)
```
Also, wouldn't the value from `next()` be lost here?

Linchin replied:
I think when `analyze == True`,  the query isn't run so there's no query results returned anyway. `explain_metrics` is parsed when calling `next()`. Are there any other returned value that I'm not aware of?

Linchin replied:
I actually copied from your PR. This is from your branch: https://github.com/googleapis/python-datastore/blob/e67ca728a982b15b6d2bc0f18fe9c27f07014471/google/cloud/datastore/query.py#L871-L872

Also updated the comments.

Linchin committed:
yield tuple instead

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -154,6 +182,10 @@ def _make_stream(
                 break
+
+            if response.explain_metrics:
+                yield response.explain_metrics
```
Good point! Done

Linchin commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -503,6 +504,48 @@ def _to_protobuf(self) -> Optional[Document]:
+
+    def __init__(
+        self,
+        docs: list,
```
I feel like `list[DocumentSnapshot]` doesn't indicate that the result has more methods than `list`, so I'm more prone to `QueryResults` or `QueryResultsList`.

daniel-sanche replied:
Sorry, I think I was talking about two different things in these comments:
- the class itself should be called either `QueryResults[T]` or `QueryResultsList[T]` (I kind of lean towards QueryResultsList, but I'm fine with either)
- the `docs: list` init argument should have a generic component to it: `docs: list[T]`

Linchin replied:
I like QueryResultsList more too! done

Linchin committed:
raise exception when explain_metrics is absent

Linchin committed:
refactor documentsnapshotlist into queryresultslist

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -54,6 +58,7 @@ def get(
         ] = gapic_v1.method.DEFAULT,
         timeout: float | None = None,
+        explain_options: Optional[ExplainOptions] = None,
     ) -> List[AggregationResult]:
```
done

Linchin committed:
add comment

Linchin committed:
improve type hint

Linchin commented on google/cloud/firestore_v1/collection.py:
```diff
@@ -169,7 +171,8 @@ def get(
         timeout: Union[float, None] = None,
-    ) -> list:
+        explain_options: Optional[ExplainOptions] = None,
+    ) -> DocumentSnapshotList:
```
indeed, done

Linchin committed:
lint

daniel-sanche commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -503,6 +522,45 @@ def _to_protobuf(self) -> Optional[Document]:
         return _helpers.document_snapshot_to_protobuf(self)
 
 
+class QueryResultsList(list):
```
maybe we should move this in with stream_generator.py? Or it's own file?

It made sense to leave this here when it only held DocumentReferences, but not so much now that it's generic

Linchin replied:
True! moved

Linchin replied:
I moved to its own file instead.

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -127,16 +152,21 @@ def _make_stream(
 
         Yields:
-            :class:`~google.cloud.firestore_v1.base_aggregation.AggregationResult`:
+            Tuple[Optional[DocumentSnapshot], Optional[ExplainMetrics]]:
```
shouldn't `Optional[DocumentSnapshot]` be `List[AggregationResult]`?

Linchin replied:
indeed, updated

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -154,14 +184,20 @@ def _make_stream(
                 break
+
+            if response.explain_metrics:
+                yield None, response.explain_metrics
```
no need to yield here: you can yield them both together a couple lines later. That way, only the second value needs to be optional

```
maybe_metrics = response.explain_metrics or None
result = _query_response_to_result(response)
yield result, maybe_metrics
```

Linchin replied:
I like the idea of returning the value instead! This way we can avoid calling next() internally too.

daniel-sanche commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -230,6 +234,8 @@ def get(
         timeout: float | None = None,
+        *,
+        explain_options: Optional[ExplainOptions] = None,
     ) -> List[AggregationResult] | Coroutine[Any, Any, List[AggregationResult]]:
```
update the return type? (looks like there are a lot that can be changed across the files)

Linchin replied:
done, also checked the whole file

daniel-sanche commented on google/cloud/firestore_v1/transaction.py:
```diff
@@ -171,6 +174,8 @@ def get(
         timeout: float = None,
+        *,
+        explain_options: Optional[ExplainOptions] = None,
     ) -> Generator[DocumentSnapshot, Any, None]:
```
change type?

Linchin replied:
done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -50,10 +49,14 @@ def __iter__(self):
+
+            # Need to run the following iteration too, to ensure the length of
+            # the iteration isn't increased due to yielding explain_metrics.
+            return next(self)
```
We can simplify this if we change the stream to yield both values together, right?

Linchin committed:
move QueryResultsList to stream_generator.py

Linchin committed:
aggregation related type annotation

Linchin committed:
transaction return type hint

Linchin committed:
refactor QueryResultsList

Linchin committed:
change stream generator to return ExplainMetrics instead of yield

Linchin committed:
update aggregation query to use the new generator

Linchin committed:
update query to use the new generator

Linchin committed:
update vector query to use the new generator

Linchin committed:
lint

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -14,21 +14,54 @@
-    """Generator for the streamed results."""
+    """Generator for the streamed results.
 
-    def __init__(self, response_generator):
```
FYI: https://github.com/googleapis/python-firestore/issues/947

Maybe we should look at this as part of the changes?

Linchin replied:
Thanks, I think with the mypy change this should be resolved, although there are still many type annotation errors remaining.

Linchin committed:
type annotations

Linchin committed:
Merge branch 'main' into query-profiling-3

Linchin committed:
fix type annotation to be python 3.9 compatible

Linchin committed:
fix type hint for python 3.8

Linchin committed:
fix system test

Linchin committed:
add test coverage

Linchin committed:
use class method get_explain_metrics() instead of property explain_metrics

daniel-sanche commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -62,7 +66,7 @@ class AggregationResult(object):
     """
 
-    def __init__(self, alias: str, value: int, read_time=None):
+    def __init__(self, alias: str, value: int | float, read_time=None):
```
You can just use `float` instead of `int | float` (float implies int)

Linchin replied:
That's good to know! Replaced all occurrences.

daniel-sanche commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -230,7 +237,12 @@ def get(
+        explain_options: Optional[ExplainOptions] = None,
+    ) -> (
+        QueryResultsList[AggregationResult]
+        | Coroutine[Any, Any, List[AggregationResult]]
```
Shouldn't the Coroutine also return QueryResultsList? (Also, `Awaitable[` seems cleaner than `Coroutine[Any, Any`

Linchin replied:
I didn't change this part because async isn't implemented in this PR. I will change it to use `Awaitable[` instead.

daniel-sanche commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -246,21 +258,30 @@ def get(
-        Generator[List[AggregationResult], Any, None]
-        | AsyncGenerator[List[AggregationResult], None]
+        StreamGenerator[Optional[List[AggregationResult]]]
+        | AsyncStreamGenerator[List[AggregationResult], Any, None]
```
does AsyncStreamGenerator exist in this PR?

Linchin replied:
I think `stream()` does return AsyncStreamGenerator, [see here](https://github.com/googleapis/python-firestore/blob/ea64f23a68d7610ce7ed3369c723e2cbffd2c56d/google/cloud/firestore_v1/async_aggregation.py#L164).

Linchin replied:
But I removed the async return value :)

daniel-sanche commented on google/cloud/firestore_v1/base_document.py:
```diff
@@ -13,17 +13,33 @@
+    from google.cloud.firestore_v1.types import Document, firestore, write
+
+
+T = TypeVar("T")
```
is this needed?

Linchin replied:
Thanks for catching this. I think I added generic type here at first, but removed because sometimes it's not accepted by python 3.7.

Linchin replied:
removed

daniel-sanche commented on google/cloud/firestore_v1/base_vector_query.py:
```diff
@@ -124,5 +134,7 @@ def stream(
-    ) -> Iterable[document.DocumentSnapshot]:
+        *,
+        explain_options: Optional[ExplainOptions] = None,
+    ) -> Generator[Optional[DocumentSnapshot], Any, Optional[ExplainMetrics]]:
```
why Optional[DocumentSnapshot]?

Linchin replied:
I added the `Optional` to cover the case when the generator generates 0 items. Maybe it's not necessary?

Linchin replied:
I think it's not necessary. I removed all similar occurrences.

daniel-sanche replied:
Yeah, that's not necessary. Optional implies it would sometimes `yield None`, which would be different from just not yielding anything

daniel-sanche commented on google/cloud/firestore_v1/query.py:
```diff
@@ -203,7 +223,7 @@ def _chunkify(
 
             if snapshots:
-                last_document = snapshots[-1]
+                last_document = snapshots[-1]  # type: ignore
```
why type ignore?

Linchin replied:
Some unit test wasn't passing but somehow it works now. I'll undo the change.

daniel-sanche commented on google/cloud/firestore_v1/query.py:
```diff
@@ -349,15 +371,26 @@ def _make_stream(
 
         Yields:
-            :class:`~google.cloud.firestore_v1.document.DocumentSnapshot`:
+            Optional[DocumentSnapshot]:
```
I don't think this needs to be optional

Linchin replied:
Indeed, done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -14,21 +14,50 @@
 
-    def __init__(self, response_generator):
+
+class StreamGenerator(Generator[Any, Any, Optional[ExplainMetrics]]):
```
1. We should use a generic for the output type, instead of Any
2. We probably don't need to return Optional[ExplainMetrics] to the end user in the API contract, we can treat that as an implementation detail. Users should access explain_metrics by calling the method

Linchin replied:
I think you are right about both. I will update accordingly.

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -154,15 +193,29 @@ def _make_stream(
+                # do not yield an empty result.
+                yield result
+
+        return metrics
```
I'm realizing AsyncGenerator doesn't support returning values, so we'll have to find a different method for the async side... I don't know if that changes anything here

Linchin replied:
I encountered this issue when implementing the async part. The temp hacky solution is to yield the metrics and catch it by checking type 😂 I think I really like the return here so I'd rather the sync and async aren't symmetric here, but let's discuss in more details in our meeting.

daniel-sanche replied:
Yeah, I guess we're back to our old worse options for the Async side! I agree it's probably fine to keep them separate if it saves us headaches on the sync side at least

Linchin committed:
Merge branch 'main' into query-profiling-3

Linchin commented on google/cloud/firestore_v1/query_results.py:
```diff
@@ -0,0 +1,74 @@
+from typing import List, Optional, TypeVar
+
+
+T = TypeVar("T")
```
Good catch! removed.

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -127,16 +155,27 @@ def _make_stream(
 
         Yields:
-            :class:`~google.cloud.firestore_v1.base_aggregation.AggregationResult`:
+            Optional[List[AggregationResult]]:
```
I think you are right, even without the Optional the generator should be allowed to be empty.

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -154,15 +193,29 @@ def _make_stream(
+                # When explain_options.analyze == False, the query isn't run
+                # but query profiling results are returned. In this case we
+                # do not yield an empty result.
+                yield result
```
I think this logic was intended to cover the case when `analyze==True` but aggregation result is empty. In this case we still yield an empty list. I added more comments to explain the reasoning more clearly. I also changed the condition to be more direct, as following:
```
if len(result) > 0 or explain_options.analyze is True:
```

daniel-sanche replied:
That makes sense. What happens if we just don't yield anything though? Does it break when analyze is True?

daniel-sanche replied:
It seems like we have to decide what we want to happen when analyze is True:
- yield a single, empty list
- just raise a StopIteration, as an empty iterator

If we yield an emtpy list, that means `list(genrator) == [[]]`. I think just having no yields feels cleaner to me. WDYT?

Linchin replied:
Indeed yielding nothing didn't break any test, I'll just do this!

Linchin commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -154,15 +193,29 @@ def _make_stream(
-    ) -> "StreamGenerator[DocumentSnapshot]":
+        *,
+        explain_options: Optional[ExplainOptions] = None,
+    ) -> StreamGenerator[Optional[List[AggregationResult]]]:
```
Indeed, as you can see [here](https://github.com/googleapis/python-firestore/blob/ea64f23a68d7610ce7ed3369c723e2cbffd2c56d/google/cloud/firestore_v1/aggregation.py#L111) the internal `_make_stream()` has always yielded AggregationResult before.

Linchin committed:
address comments

Linchin committed:
remove more Optional

daniel-sanche commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -230,7 +237,12 @@ def get(
+        explain_options: Optional[ExplainOptions] = None,
+    ) -> (
+        QueryResultsList[AggregationResult]
+        | Awaitable[Any, Any, List[AggregationResult]]
```
I think Awaitable only takes one argument. And I think it should be QueryResultsList, not just List

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -14,21 +14,53 @@
-        return self._generator
+        return self
 
     def __next__(self):
```
nit: you can probably add `-> T` return annotation here, which might help with type checking

Linchin replied:
done

Linchin committed:
add type hint for async stream generator

Linchin committed:
simplify yield in aggregation stream

Linchin committed:
stream generator type annotation

daniel-sanche commented:
One last thing: I ran mypy locally, and found a couple new type issues that seem relevant to these changes

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.base_aggregation                                                                                                                                          
google/cloud/firestore_v1/base_aggregation.py:243: error: "QueryResultsList" expects no type arguments, but 1 given  [type-arg]
google/cloud/firestore_v1/base_aggregation.py:243: error: "Awaitable" expects 1 type argument, but 3 given  [type-arg]
google/cloud/firestore_v1/base_aggregation.py:283: error: "AsyncStreamGenerator" expects no type arguments, but 2 given  [type-arg]
```

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.collection                                                                                                                                                  
google/cloud/firestore_v1/collection.py:170: error: Return type "QueryResultsList" of "get" incompatible with return type "Generator[DocumentSnapshot, Any, Any] | AsyncGenerator[DocumentSnapshot, Any]" in supertype "BaseCollectionReference"  [override]
google/cloud/firestore_v1/collection.py:177: error: "QueryResultsList" expects no type arguments, but 1 given  [type-arg]
```

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.query_results
google/cloud/firestore_v1/query_results.py:71: error: Incompatible return value type (got "ExplainMetrics | None", expected "ExplainMetrics")  [return-value]
```

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.stream_generator                                                                                                                                         
google/cloud/firestore_v1/stream_generator.py:68: error: Signature of "throw" incompatible with supertype "Generator"  [override]
google/cloud/firestore_v1/stream_generator.py:68: note:      Superclass:
google/cloud/firestore_v1/stream_generator.py:68: note:          @overload
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, type[BaseException], object = ..., TracebackType | None = ..., /) -> T
google/cloud/firestore_v1/stream_generator.py:68: note:          @overload
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, BaseException, None = ..., TracebackType | None = ..., /) -> T
google/cloud/firestore_v1/stream_generator.py:68: note:      Subclass:
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, exp: Any = ...) -> Any

```

Linchin commented:
```
google/cloud/firestore_v1/base_aggregation.py:243: error: "QueryResultsList" expects no type arguments, but 1 given  [type-arg]
```
I clean up `QueryResultsList` related type hints.

```
google/cloud/firestore_v1/base_aggregation.py:243: error: "Awaitable" expects 1 type argument, but 3 given  [type-arg]
```
Now that I think about it, I don't think get() returns a coroutine here, so I changed it to the list.

```
google/cloud/firestore_v1/base_aggregation.py:283: error: "AsyncStreamGenerator" expects no type arguments, but 2 given  [type-arg]
```
Added generic yield type for AsyncStreamGenerator, but we will be using it as `AsyncStreamGenerator[YieldType]`, instead of `AsyncStreamGenerator[YieldType, SendType]`. I'm not sure if it's worth the effort to create a second generic type just for this.

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.collection                                                                                                                                                  
google/cloud/firestore_v1/collection.py:170: error: Return type "QueryResultsList" of "get" incompatible with return type "Generator[DocumentSnapshot, Any, Any] | AsyncGenerator[DocumentSnapshot, Any]" in supertype "BaseCollectionReference"  [override]
google/cloud/firestore_v1/collection.py:177: error: "QueryResultsList" expects no type arguments, but 1 given  [type-arg]
```
Cleaned this up.

```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.query_results
google/cloud/firestore_v1/query_results.py:71: error: Incompatible return value type (got "ExplainMetrics | None", expected "ExplainMetrics")  [return-value]
```
I feel like this is caused by mypy being unaware of the inner logic of the function. It either returns ExplainMetrics, or raises an exception, so there's no way it returns None. I will add `# type: ignore` to this line.


```
tmp/python-firestore [query-profiling-3●] » mypy -p google.cloud.firestore_v1.stream_generator                                                                                                                                         
google/cloud/firestore_v1/stream_generator.py:68: error: Signature of "throw" incompatible with supertype "Generator"  [override]
google/cloud/firestore_v1/stream_generator.py:68: note:      Superclass:
google/cloud/firestore_v1/stream_generator.py:68: note:          @overload
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, type[BaseException], object = ..., TracebackType | None = ..., /) -> T
google/cloud/firestore_v1/stream_generator.py:68: note:          @overload
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, BaseException, None = ..., TracebackType | None = ..., /) -> T
google/cloud/firestore_v1/stream_generator.py:68: note:      Subclass:
google/cloud/firestore_v1/stream_generator.py:68: note:          def throw(self, exp: Any = ...) -> Any
```
I'm aware of the overload incompatibility here, but if we change the method signature now, we risk being backward incompatible with the version we just released 😅

Linchin committed:
more type hints

Linchin committed:
remove "Integer"

Linchin committed:
docstring format

daniel-sanche commented on google/cloud/firestore_v1/query_results.py:
```diff
@@ -68,4 +69,4 @@ def get_explain_metrics(self) -> ExplainMetrics:
             raise QueryExplainError("explain_options not set on query.")
         else:
-            return self._explain_metrics
+            return self._explain_metrics  # type: ignore
```
> I feel like this is caused by mypy being unaware of the inner logic of the function. It either returns ExplainMetrics, or raises an exception, so there's no way it returns None. I will add # type: ignore to this line.

I don't think that's true. There's nothing stopping someone from passing explain_options, but not explain_metrics:
```
instance = QueryResultList([], object(), None)
instance.get_explain_metrics() == None
```

There are a couple ways we can address this. Maybe we want to ensure that if one is set, they both are, and store them together as a tuple? Or maybe we should just add a separate exception here if self._explain_metrics is None as well? I can't remember off the top of my head if it makes sense to have one but not the other here

---

Either way, I think adding `# type: ignore` should be a last resort. We can sometimes use `typing.cast` if mypy ever has trouble following the invariants. But I think in this case, it was actually a good catch in the logic

Linchin replied:
You are right, mypy did have a good point! When explain_options is set, ultimately server will return explain_metrics, and it should already been gathered before QueryResultList is initialized. To make QueryResultsList more robust, I can do the following:
1. When initializing, throw an error when `explain_options is not None and explain_metrics is None`. (maybe even the case `explain_options is None and explain_metrics is not None`?)
2. Add a check in `get_explain_metrics()` like the following in case `self._explain_metrics` got wiped out somewhere
```
    def get_explain_metrics(self) -> ExplainMetrics:
        if self._explain_options is None:
            raise QueryExplainError("explain_options not set on query.")
        elif self._explain_metrics is None:
            raise QueryExplainError("explain_metrics is empty despite explain_options is set.")
        else:
            return self._explain_metrics
```

daniel-sanche replied:
That works, or you can require them to be passed as a tuple, which would solve the same typing problem with less checks:

```
class QueryResultsList:

    def __init__(
        self,
        docs: List,
        profile_data: tuple[ExplainOptions, ExplainMetrics] | None = None
    ):
```

Linchin replied:
done

daniel-sanche commented on google/cloud/firestore_v1/stream_generator.py:
```diff
@@ -30,7 +30,7 @@
 
 
-class StreamGenerator(Generator[T, Any, None]):
+class StreamGenerator(Generator[T, Any, Optional[ExplainMetrics]]):
```
> I'm aware of the overload incompatibility here, but if we change the method signature now, we risk being backward incompatible with the version we just released 😅

It looks like this passes the check:

```
def throw(self, exp=None, value=None, traceback=None) -> T:
    return self._generator.throw(exp, value, traceback)
```

This should be backwards compatible, since we're just adding new optional arguments. All existing code should be unaffected

---

Honestly though, I think removing the `=None` from `exp=None` would be best. exp is meant to be an exception, so passing None just raises a TypeError. Usually we should avoid potential breaking changes, but I don't think that applies in this case, where the default itself broken. Especially since this was such a recent addition. Let me know what you think

Linchin replied:
You are right, we shouldn't let `exp` be optional. Just to track our discussion offline, I tried to add type annotations for all the parameters here, but it seems mypy has trouble matching it with the annotations that are present in the parent class `Generator`. So we will skip the annotation here and let mypy infer it from the parent class.

Linchin replied:
Additionally, python 3.12 has [deprecated](https://docs.python.org/3/reference/expressions.html#generator.throw) parameters `value` and `traceback` in `generator.throw()`, so to be compatible with both 3.12 and older versions, we will just use `throw(*args, **kwargs)`.

Linchin commented on google/cloud/firestore_v1/base_aggregation.py:
```diff
@@ -264,9 +253,8 @@ def get(
 
         Returns:
-            QueryResultsList[AggregationResult] | Awaitable[Any, Any, List[AggregationResult]]:
+            (QueryResultsList[List[AggregationResult]] | List[List[AggregationResult]]):
```
Sounds good, but after trying it out I think both`Coroutine[Any, Any, List[List[AggregationResult]]]` and `Awaitable[List[List[AggregationResult]]]` work? (maybe I'm missing something)

daniel-sanche replied:
This is what I see:

```
google/cloud/firestore_v1/async_aggregation.py:49: error: Signature of "get" incompatible with supertype "BaseAggregationQuery"  [override]
google/cloud/firestore_v1/async_aggregation.py:49: note:      Superclass:
google/cloud/firestore_v1/async_aggregation.py:49: note:          def get(self, transaction: Any = ..., retry: Retry | _MethodDefault | None = ..., timeout: float | None = ..., *, explain_options: ExplainOptions | None = ...) -> QueryResultsList[list[AggregationResult]] | Coroutine[Any, Any, list[AggregationResult]]
google/cloud/firestore_v1/async_aggregation.py:49: note:      Subclass:
google/cloud/firestore_v1/async_aggregation.py:49: note:          def get(self, transaction: Any = ..., retry: AsyncRetry | _MethodDefault | None = ..., timeout: float | None = ...) -> Coroutine[Any, Any, list[AggregationResult]]
```

But if you found it works for you, that's fine with me too! Either way, I think this can fall into the known-broken mypy issues to fix later

Linchin replied:
I see, I will revert the change and save this for later haha

Linchin committed:
mypy

Linchin committed:
add more input verification for query_results.py

Linchin committed:
Merge branch 'main' into query-profiling-3
